File: jbake-core/src/test/java/org/jbake/app/AssetTest.java
package org.jbake.app;

import ch.qos.logback.classic.spi.LoggingEvent;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.filefilter.FileFilterUtils;
import org.jbake.TestUtils;
import org.jbake.app.configuration.ConfigUtil;
import org.jbake.app.configuration.DefaultJBakeConfiguration;
import org.jbake.app.configuration.JBakeProperty;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junitpioneer.jupiter.TempDirectory;
import org.junitpioneer.jupiter.TempDirectory.TempDir;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.net.URL;
import java.nio.file.Path;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;


@ExtendWith(TempDirectory.class)
public class AssetTest extends LoggingTest {

    public Path folder;
    private DefaultJBakeConfiguration config;
    private File fixtureDir;


    @BeforeEach
    public void setup(@TempDir Path folder) throws Exception {
        fixtureDir = new File(this.getClass().getResource("/fixture").getFile());
        this.folder = folder;
        config = (DefaultJBakeConfiguration) new ConfigUtil().loadConfig(fixtureDir);
        config.setDestinationFolder(folder.toFile());
        Assertions.assertEquals(".html", config.getOutputExtension());
    }


    @Test
    public void testCopy() throws Exception {
        Asset asset = new Asset(config);
        asset.copy();
        File cssFile = new File(folder.toString() + File.separatorChar + "css" + File.separatorChar + "bootstrap.min.css");
        Assertions.assertTrue(cssFile.exists(), () -> "File " + cssFile.getAbsolutePath() + " does not exist");
        File imgFile = new File(folder.toString() + File.separatorChar + "img" + File.separatorChar + "glyphicons-halflings.png");
        Assertions.assertTrue(imgFile.exists(), () -> "File " + imgFile.getAbsolutePath() + " does not exist");
        File jsFile = new File(folder.toString() + File.separatorChar + "js" + File.separatorChar + "bootstrap.min.js");
        Assertions.assertTrue(jsFile.exists(), () -> "File " + jsFile.getAbsolutePath() + " does not exist");

        Assertions.assertTrue(asset.getErrors().isEmpty(), "Errors during asset copying");
    }

    @Test
    public void testCopySingleFile() throws Exception {
        Asset asset = new Asset(config);
        String cssSubPath = File.separatorChar + "css" + File.separatorChar + "bootstrap.min.css";
        String contentImgPath = File.separatorChar + "blog" + File.separatorChar + "2013" + File.separatorChar
            + "images" + File.separatorChar + "custom-image.jpg";

        // Copy single Asset File
        File expected = new File(folder.toString() + cssSubPath);
        Assertions.assertFalse(expected.exists(), "cssFile should not exist before running the test; avoids false positives");
        File cssFile = new File(fixtureDir.getPath() + File.separatorChar + "assets" + cssSubPath);
        asset.copySingleFile(cssFile);
        Assertions.assertTrue(expected.exists(), "Css asset file did not copy");

        // Copy single Content file
        expected = new File(folder.toString() + contentImgPath);
        Assertions.assertFalse(expected.exists(), "content image file should not exist before running the test");
        File imgFile = new File(fixtureDir.getPath() + File.separatorChar + "content" + contentImgPath);
        asset.copySingleFile(imgFile);
        Assertions.assertTrue(expected.exists(), "Content img file did not copy");
    }

    @Test
    public void shouldSkipCopyingSingleFileIfDirectory() throws IOException {

        Asset asset = new Asset(config);

        File emptyDir = new File(folder.toFile(),"emptyDir");
        emptyDir.mkdir();
        File expectedDir = new File(fixtureDir.getCanonicalPath(), "emptyDir");

        asset.copySingleFile(emptyDir);

        Assertions.assertFalse(expectedDir.exists(), "Directory should be skipped");
    }

    @Test
    public void shouldLogSkipCopyingSingleFileIfDirectory() throws IOException {

        Asset asset = new Asset(config);
        File emptyDir = new File(folder.toFile(),"emptyDir");
        emptyDir.mkdir();

        asset.copySingleFile(emptyDir);

        verify(mockAppender, times(1)).doAppend(captorLoggingEvent.capture());

        LoggingEvent loggingEvent = captorLoggingEvent.getValue();
        assertThat(loggingEvent.getMessage()).isEqualTo("Skip copying single asset file [{}]. Is a directory.");

    }

    @Test
    public void testCopyCustomFolder() throws Exception {
        config.setAssetFolder(new File(config.getSourceFolder(), "/media"));
        Asset asset = new Asset(config);
        asset.copy();

        File favFile = new File(folder.toString() + File.separatorChar + "favicon.ico");
        Assertions.assertTrue(favFile.exists(), () -> "File " + favFile.getAbsolutePath() + " does not exist");

        Assertions.assertTrue(asset.getErrors().isEmpty(), "Errors during asset copying");
    }

    @Test
    public void testCopyIgnore() throws Exception {
        File assetFolder = new File(folder.toFile(), "ignoredAssets");
        assetFolder.mkdirs();
        FileUtils.copyDirectory(new File(this.getClass().getResource("/fixture/ignorables").getFile()), assetFolder);
        config.setAssetFolder(assetFolder);
        config.setAssetIgnoreHidden(true);
        TestUtils.hideAssets(assetFolder);
        Asset asset = new Asset(config);
        asset.copy(assetFolder);

        File testFile = new File(folder.toFile(), "test.txt");
        Assertions.assertTrue(testFile.exists(), () -> "File " + testFile.getAbsolutePath() + " does not exist");
        File testIgnoreFile = new File(folder.toFile(), ".test.txt");
        Assertions.assertFalse(testIgnoreFile.exists(), () -> "File " + testIgnoreFile.getAbsolutePath() + " does exist");

        Assertions.assertTrue(asset.getErrors().isEmpty(), "Errors during asset copying");
    }


    /**
     * Primary intention is to extend test cases to increase coverage.
     *
     * @throws Exception
     */
    @Test
    public void testWriteProtected() throws Exception {
        File assets = new File(config.getSourceFolder(), "assets");
        File css = new File(folder.toFile(),"css");
        css.mkdir();
        final File cssFile = new File(css, "bootstrap.min.css");
        FileUtils.touch(cssFile);
        cssFile.setReadOnly();

        config.setAssetFolder(assets);
        config.setDestinationFolder(folder.toFile());
        Asset asset = new Asset(config);
        asset.copy();

        cssFile.setWritable(true);
        Assertions.assertFalse(asset.getErrors().isEmpty(), "At least one error during copy expected");
    }

    /**
     * Primary intention is to extend test cases to increase coverage.
     *
     * @throws Exception
     */
    @Test
    public void testUnlistable() throws Exception {
        config.setAssetFolder(new File(config.getSourceFolder(), "non-exsitent"));
        Asset asset = new Asset(config);
        asset.copy();
    }

    @Test
    public void testJBakeIgnoredFolder() {
        URL assetsUrl = this.getClass().getResource("/fixture/assets");
        File assets = new File(assetsUrl.getFile());
        Asset asset = new Asset(config);
        asset.copy(assets);

        File cssFile = new File(folder.toString() + File.separatorChar + "css" + File.separatorChar + "bootstrap.min.css");
        Assertions.assertTrue(cssFile.exists(), () -> "File " + cssFile.getAbsolutePath() + " does not exist");
        File imgFile = new File(folder.toString() + File.separatorChar + "img" + File.separatorChar + "glyphicons-halflings.png");
        Assertions.assertTrue(imgFile.exists(), () -> "File " + imgFile.getAbsolutePath() + " does not exist");
        File jsFile = new File(folder.toString() + File.separatorChar + "js" + File.separatorChar + "bootstrap.min.js");
        Assertions.assertTrue(jsFile.exists(), () -> "File " + jsFile.getAbsolutePath() + " does not exist");
        File ignorableFolder = new File(folder.toString() + File.separatorChar + "ignorablefolder");
        File fooIgnorableFolder = new File(folder.toString() + File.separatorChar + "fooignorablefolder");
        Assertions.assertFalse(ignorableFolder.exists(), () -> "Folder " + ignorableFolder.getAbsolutePath() + " must not exist");
        Assertions.assertTrue(fooIgnorableFolder.exists(), () -> "Folder " + fooIgnorableFolder.getAbsolutePath() + " must exist");

        Assertions.assertTrue(asset.getErrors().isEmpty(), "Errors during asset copying");
    }

    @Test
    public void testFooIgnoredFolder() {
        config.setProperty(JBakeProperty.IGNORE_FILE, ".fooignore");

        URL assetsUrl = this.getClass().getResource("/fixture/assets");
        File assets = new File(assetsUrl.getFile());
        Asset asset = new Asset(config);
        asset.copy(assets);

        File cssFile = new File(folder.toString() + File.separatorChar + "css" + File.separatorChar + "bootstrap.min.css");
        Assertions.assertTrue(cssFile.exists(), () -> "File " + cssFile.getAbsolutePath() + " does not exist");
        File imgFile = new File(folder.toString() + File.separatorChar + "img" + File.separatorChar + "glyphicons-halflings.png");
        Assertions.assertTrue(imgFile.exists(), () -> "File " + imgFile.getAbsolutePath() + " does not exist");
        File jsFile = new File(folder.toString() + File.separatorChar + "js" + File.separatorChar + "bootstrap.min.js");
        Assertions.assertTrue(jsFile.exists(), () -> "File " + jsFile.getAbsolutePath() + " does not exist");
        File ignorableFolder = new File(folder.toString() + File.separatorChar + "ignorablefolder");
        File fooIgnorableFolder = new File(folder.toString() + File.separatorChar + "fooignorablefolder");
        Assertions.assertTrue(ignorableFolder.exists(), () -> "Folder " + ignorableFolder.getAbsolutePath() + " must exist");
        Assertions.assertFalse(fooIgnorableFolder.exists(), () -> "Folder " + fooIgnorableFolder.getAbsolutePath() + " must not exist");

        Assertions.assertTrue(asset.getErrors().isEmpty(), "Errors during asset copying");
    }

    @Test
    public void testCopyAssetsFromContent() {
        URL contentUrl = this.getClass().getResource("/fixture/content");
        File contents = new File(contentUrl.getFile());
        Asset asset = new Asset(config);
        asset.copyAssetsFromContent(contents);

        int totalFiles = countFiles(folder.toFile());
        int expected = 3;

        Assertions.assertTrue(totalFiles == expected, () -> String.format("Number of files copied must be %d but are %d", expected, totalFiles));

        File pngFile = new File(folder.toString() + File.separatorChar + "blog" + File.separatorChar + "2012/images/custom-image.png");
        Assertions.assertTrue(pngFile.exists(), () -> "File " + pngFile.getAbsolutePath() + " does not exist");

        File jpgFile = new File(folder.toString() + File.separatorChar + "blog" + File.separatorChar + "2013/images/custom-image.jpg");
        Assertions.assertTrue(jpgFile.exists(), () -> "File " + jpgFile.getAbsolutePath() + " does not exist");

        File jsonFile = new File(folder.toString() + File.separatorChar + "blog" + File.separatorChar + "2012/sample.json");
        Assertions.assertTrue(jsonFile.exists(), () -> "File " + jsonFile.getAbsolutePath() + " does not exist");

        Assertions.assertTrue(asset.getErrors().isEmpty(), "Errors during asset copying");
    }

    @Test
    public void testIsFileAsset() {
        File cssAsset = new File(config.getAssetFolder().getAbsolutePath() + File.separatorChar + "css" + File.separatorChar + "bootstrap.min.css");
        Assertions.assertTrue(cssAsset.exists());
        File contentFile = new File(config.getContentFolder().getAbsolutePath() + File.separatorChar + "about.html");
        Assertions.assertTrue(contentFile.exists());
        Asset asset = new Asset(config);

        Assertions.assertTrue(asset.isAssetFile(cssAsset));
        Assertions.assertFalse(asset.isAssetFile(contentFile));
    }


    private Integer countFiles(File path) {
        int total = 0;
        FileFilter filesOnly = FileFilterUtils.fileFileFilter();
        FileFilter dirsOnly = FileFilterUtils.directoryFileFilter();
        File[] files = path.listFiles(filesOnly);
        System.out.println(files);
        total += files.length;
        for (File file : path.listFiles(dirsOnly)) {
            total += countFiles(file);
        }
        return total;
    }
}

File: jbake-core/src/test/java/org/jbake/app/AssetTest.kt
package org.jbake.app

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.apache.commons.io.FileUtils
import org.apache.commons.io.filefilter.FileFilterUtils
import org.jbake.TestUtils
import org.jbake.app.configuration.ConfigUtil
import org.jbake.app.configuration.DefaultJBakeConfiguration
import org.jbake.app.configuration.PropertyList
import java.io.File
import java.nio.file.Files

class AssetTest : StringSpec({

    lateinit var tempDir: File
    lateinit var config: DefaultJBakeConfiguration
    lateinit var fixtureDir: File

    beforeTest {
        fixtureDir = File(AssetTest::class.java.getResource("/fixture")!!.file)
        tempDir = Files.createTempDirectory("jbake-test").toFile()
        config = ConfigUtil().loadConfig(fixtureDir) as DefaultJBakeConfiguration
        config.destinationDir = tempDir
        config.outputExtension shouldBe ".html"
    }

    afterTest {
        tempDir.deleteRecursively()
    }

    "testCopy" {
        val asset = Asset(config)
        asset.copy()

        tempDir.resolve("css/bootstrap.min.css").exists() shouldBe true
        tempDir.resolve("img/glyphicons-halflings.png").exists() shouldBe true
        tempDir.resolve("js/bootstrap.min.js").exists() shouldBe true
        asset.errors.isEmpty() shouldBe true
    }

    "testCopySingleFile" {
        val asset = Asset(config)

        // Copy single asset file
        val cssTarget = tempDir.resolve("css/bootstrap.min.css")
        cssTarget.exists() shouldBe false
        asset.copySingleFile(File(fixtureDir, "assets/css/bootstrap.min.css"))
        cssTarget.exists() shouldBe true

        // Copy single content file
        val imgTarget = tempDir.resolve("blog/2013/images/custom-image.jpg")
        imgTarget.exists() shouldBe false
        asset.copySingleFile(File(fixtureDir, "content/blog/2013/images/custom-image.jpg"))
        imgTarget.exists() shouldBe true
    }

    "shouldSkipCopyingSingleFileIfDirectory" {
        val asset = Asset(config)
        val emptyDir = tempDir.resolve("emptyDir").apply { mkdir() }

        asset.copySingleFile(emptyDir)

        File(fixtureDir.canonicalPath, "emptyDir").exists() shouldBe false
    }

    "shouldLogSkipCopyingSingleFileIfDirectory" {
        // This test requires LoggingTest setup - skipping for now or implement MockK logging
        val asset = Asset(config)
        tempDir.resolve("emptyDir").apply { mkdir() }.let { asset.copySingleFile(it) }
        // TODO: verify logging with MockK appender
    }

    "testCopyCustomDir" {
        config.assetDir = config.sourceDir.resolve("media")
        Asset(config).copy()

        tempDir.resolve("favicon.ico").exists() shouldBe true
    }

    "testCopyIgnore" {
        val assetDir = tempDir.resolve("ignoredAssets").apply { mkdirs() }
        FileUtils.copyDirectory(File(AssetTest::class.java.getResource("/fixture/ignorables")!!.file), assetDir)
        config.assetDir = assetDir
        config.assetIgnoreHidden = true
        TestUtils.hideAssets(assetDir)

        val asset = Asset(config)
        asset.copy(assetDir)

        tempDir.resolve("test.txt").exists() shouldBe true
        tempDir.resolve(".test.txt").exists() shouldBe false
        asset.errors.isEmpty() shouldBe true
    }

    "testWriteProtected" {
        val css = tempDir.resolve("css").apply { mkdir() }
        val cssFile = File(css, "bootstrap.min.css").apply {
            FileUtils.touch(this)
            setReadOnly()
        }
        css.setReadOnly()

        config.assetDir = config.sourceDir.resolve("assets")
        config.destinationDir = tempDir
        val asset = Asset(config)
        asset.copy()

        css.setWritable(true) // Restore directory permissions first
        cssFile.setWritable(true)
        asset.errors.isEmpty() shouldBe false
    }

    "testUnlistable" {
        config.assetDir = config.sourceDir.resolve("non-existent")
        Asset(config).copy()
    }

    "testJBakeIgnoredDir" {
        val assets = File(AssetTest::class.java.getResource("/fixture/assets")!!.file)
        val asset = Asset(config)
        asset.copy(assets)

        tempDir.resolve("css/bootstrap.min.css").exists() shouldBe true
        tempDir.resolve("img/glyphicons-halflings.png").exists() shouldBe true
        tempDir.resolve("js/bootstrap.min.js").exists() shouldBe true
        tempDir.resolve("ignorablefolder").exists() shouldBe false
        tempDir.resolve("fooignorablefolder").exists() shouldBe true
        asset.errors.isEmpty() shouldBe true
    }

    "testFooIgnoredDir" {
        config.setProperty(PropertyList.IGNORE_FILE.key, ".fooignore")
        val assets = File(AssetTest::class.java.getResource("/fixture/assets")!!.file)
        val asset = Asset(config)
        asset.copy(assets)

        tempDir.resolve("css/bootstrap.min.css").exists() shouldBe true
        tempDir.resolve("img/glyphicons-halflings.png").exists() shouldBe true
        tempDir.resolve("js/bootstrap.min.js").exists() shouldBe true
        tempDir.resolve("ignorablefolder").exists() shouldBe true
        tempDir.resolve("fooignorablefolder").exists() shouldBe false
        asset.errors.isEmpty() shouldBe true
    }

    "testCopyAssetsFromContent" {
        val contents = File(AssetTest::class.java.getResource("/fixture/content")!!.file)
        val asset = Asset(config)
        asset.copyAssetsFromContent(contents)

        countFiles(tempDir) shouldBe 3
        tempDir.resolve("blog/2012/images/custom-image.png").exists() shouldBe true
        tempDir.resolve("blog/2013/images/custom-image.jpg").exists() shouldBe true
        tempDir.resolve("blog/2012/sample.json").exists() shouldBe true
        asset.errors.isEmpty() shouldBe true
    }

    "testIsFileAsset" {
        val cssAsset = config.assetDir.resolve("css/bootstrap.min.css")
        val contentFile = config.contentDir.resolve("about.html")
        val asset = Asset(config)

        cssAsset.exists() shouldBe true
        contentFile.exists() shouldBe true
        asset.isAssetFile(cssAsset) shouldBe true
        asset.isAssetFile(contentFile) shouldBe false
    }
}) {
    companion object {
        private fun countFiles(path: File): Int {
            val files = path.listFiles(FileFilterUtils.fileFileFilter() as java.io.FileFilter) ?: return 0
            val dirs = path.listFiles(FileFilterUtils.directoryFileFilter() as java.io.FileFilter) ?: emptyArray()
            return files.size + dirs.sumOf { countFiles(it) }
        }
    }
}
